### ISSUE TITLE
.gcno files not consistent with GCC output

### ISSUE THREAD

**Issue by stbrowne** [ccache/ccache#674]
### How to reproduce ###
No minimal sample yet. A simple sample such as the following works as expected:
```
$ ccache g++ -x c++ -fprofile-dir=. -fprofile-arcs -ftest-coverage - -o Coverage/test <<<"int main() { return 0; }"
$ ls *.gcno
#source#-.gcno
```

### Actual behavior ###
A larger code base produces unexpected results where sometimes the files are in the output/object dir. Or sometimes a subset of the gcno files are in the output dir and both are in the working dir etc.

Examples:
project 1:
```
ccache --clean - out_dir/project_src_name.gcno
ccache (from cache) - out_dir/project_src_name.gcno
g++ - working_dir/#path#to#out_dir#src_name.gcno
```

project 2:
```
ccache --clean - both paths (25 files in working_dir, 6 files in out_dir all dups)
ccache (from cache) - both paths (25 files in working_dir, 19 files in out_dir all dups)
g++ - working_dir/#path#to#out_dir#project_src_name.gcno
```

### Expected behavior ###
All tests should produce results consistent with running g++ directly (ie in the working_dir).

### Environment ###
ccache version 3.7.11
g++ 10.2.0 - Manually compiled
```
$ g++ -v 
Using built-in specs.
COLLECT_GCC=g++
COLLECT_LTO_WRAPPER=/opt/cross/2.0/x64/libexec/gcc/x86_64-linux-gnu/10.2.0/lto-wrapper
Target: x86_64-linux-gnu
Configured with: ../gcc-10.2.0/configure --prefix=/opt/cross/2.0/x64 --disable-multilib --target=x86_64-linux-gnu --enable-languages=c,c++ --disable-libsanitizer
Thread model: posix
Supported LTO compression algorithms: zlib
gcc version 10.2.0 (GCC) 
```

Since I don't have a good minimal sample yet I'm more than happy to provide any additional information or test any patches etc.

**Comment by jrosdahl** [comment]
Thanks for the bug report.

The issue you see is because GCC 9 and newer apparently have changed behavior for `-ftest-coverage` in combination with `-fprofile-dir`. Here is a script that reproduces the behavior:

```sh
#!/bin/sh

set -eu

export CCACHE_DISABLE=1
cc=${1:-gcc}

tmpdir=$(mktemp -d)
trap "rm -rf $tmpdir" EXIT
cd $tmpdir

compile() {
    touch test.c
    $cc "$@" -c test.c 2>/dev/null
    printf "%-60s -> " "$cc $* -c test.c"
    for name in test.gcno "$(echo $tmpdir | sed 's!/!#!g')#test.gcno"; do
        if [ -e "$name" ]; then
            echo "$name"
        fi
    done
    rm -f *test*
}

mkdir dir
compile -ftest-coverage
compile -fprofile-dir=dir -ftest-coverage
```

Output when running with the compilers I have available:

```console
% for cc in gcc-4.7 gcc-4.9 gcc-7 gcc-8 gcc-9 clang-3.6 clang-3.8 clang-6.0 clang-7 clang-8 clang-9 clang-10; do ./repro $cc; done
gcc-4.7 -ftest-coverage -c test.c                            -> test.gcno
gcc-4.7 -fprofile-dir=dir -ftest-coverage -c test.c          -> test.gcno
gcc-4.9 -ftest-coverage -c test.c                            -> test.gcno
gcc-4.9 -fprofile-dir=dir -ftest-coverage -c test.c          -> test.gcno
gcc-7 -ftest-coverage -c test.c                              -> test.gcno
gcc-7 -fprofile-dir=dir -ftest-coverage -c test.c            -> test.gcno
gcc-8 -ftest-coverage -c test.c                              -> test.gcno
gcc-8 -fprofile-dir=dir -ftest-coverage -c test.c            -> test.gcno
gcc-9 -ftest-coverage -c test.c                              -> test.gcno
gcc-9 -fprofile-dir=dir -ftest-coverage -c test.c            -> #tmp#tmp.xHW5kiOag1#test.gcno
clang-3.6 -ftest-coverage -c test.c                          -> test.gcno
clang-3.6 -fprofile-dir=dir -ftest-coverage -c test.c        -> test.gcno
clang-3.8 -ftest-coverage -c test.c                          -> test.gcno
clang-3.8 -fprofile-dir=dir -ftest-coverage -c test.c        -> test.gcno
clang-6.0 -ftest-coverage -c test.c                          -> test.gcno
clang-6.0 -fprofile-dir=dir -ftest-coverage -c test.c        -> test.gcno
clang-7 -ftest-coverage -c test.c                            -> test.gcno
clang-7 -fprofile-dir=dir -ftest-coverage -c test.c          -> test.gcno
clang-8 -ftest-coverage -c test.c                            -> test.gcno
clang-8 -fprofile-dir=dir -ftest-coverage -c test.c          -> test.gcno
clang-9 -ftest-coverage -c test.c                            -> test.gcno
clang-9 -fprofile-dir=dir -ftest-coverage -c test.c          -> test.gcno
clang-10 -ftest-coverage -c test.c                           -> test.gcno
clang-10 -fprofile-dir=dir -ftest-coverage -c test.c         -> test.gcno
```

Ccache was actually updated in version 3.7.10 to handle `-fprofile-dir` in combination in with `-fprofile-arcs` and friends for GCC 9+ (#582, c9fdcef576dd) but I didn't think about checking if the coverage files were affected as well since the GCC documentation doesn't mention them. But regardless of that, ccache 3.7.x cannot know which filename to use for the `.gcno` file since it doesn't know which compiler is being used. We'll have to simply give up (i.e., fall back to running the compiler and not caching) if the `.gcno` file is not written to the expected place.

It will however be possible to handle the GCC 9+ `.gcno` files in ccache 4.x since the result files can include information about the original `.gcno` filename.

**Comment by jrosdahl** [comment]
Fixed in ccache 3.7.12. Please try it out!

**Comment by stbrowne** [comment]
Thanks for the quick turn around time. I can confirm it works in 3.7.12 now. I also removed the -fprofile-dir parameter to verify that caching works if that's gone and it does. I guess I'll probably want to re-work a few things and not use that parameter so we can keep benefiting from the ccache performance.
